### PR TITLE
tau/jet up/down variations in 2017 skimmer

### PR DESCRIPTION
**Pull request by camendola**
added:
* bjets pt, mass (jet up/down)
* taus pt, mass (tes up/down, ees up/down)
* bH pt, mass (jet up/down)
* HH pt (jet up/down, tes up/down, ees up/down)

commented out:
* Kinfit and MET up/down variations

### CODE REVIEW

**Comment by camendola** [comment]
added: 
* VBF jets pt, e (jet up/down)
* VBF jj mass (jet up/down)

removed: 
* obsolete TT mass branches

**Comment by camendola** [comment]
added:
* JER up/down
* MET jet up/down

removed: 
* obsolete filelists

**Comment by camendola** [comment]
added:
* MET tes/ees shifts
* HT20 jet shifts

removed: 
* some obsolete branches in smallTree

**Comment by camendola** [comment]
General:
* the up/down variations of the main objects are stored in vectors (of size 11 for jes, 4 for tes and 2 for ees), eg: 
```
bjet1_pt_jetup = {bjet1_pt_jetup_source1, ..., bjet1_pt_jetup_source11}
dau1_pt_tauup = {dau1_pt_tauup_DM0, dau1_pt_tauup_DM1, dau1_pt_tauup_DM10, dau1_pt_tauup_DM11}
dau1_pt_eleup = {dau1_pt_eleup_DM0, dau1_pt_eleup_DM1}
```
etc.
* the size of the vectors is also stored in a TH1F called h_syst
* for the variables that need to be read by the plotter, the variables are also stored in flat variables, eg:
```
bjet1_pt_jetup1
...
bjet1_pt_jetup11
```
* the MES up/dw variations are not splitted by DM; therefore, they are only stored in flat branches
* some code for KinFit, MT2 variations is only commented out, can be cleaned once all is implemented in the other module

File: test/skimNtuple2017.cpp [hunk 1]
```diff
@@ -461,6 +461,35 @@ double getContentHisto3D(TH3* histo, double x, double y, double z)
...
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_HF_up ->at (idx));                      //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_BBEC1_up ->at (idx));                   //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_EC2_up ->at (idx));                     //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_Absolute_up ->at (idx));                //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_BBEC1_YEAR_up ->at (idx));              //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_EC2_YEAR_up ->at (idx));                //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_Absolute_YEAR_up ->at (idx));           //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_HF_YEAR_up ->at (idx));                 //  ...                  , up
+  unc_up.push_back(theBigTree.jets_jetUncRegrouped_RelativeSample_YEAR_up ->at (idx));     //  uncertainty source 11, up
+
```
Comment (camendola) [inline test/skimNtuple2017.cpp:713]: @francescobrivio can you cross-check that I picked the right uncertainties?

Comment (francescobrivio) [reply test/skimNtuple2017.cpp:713]: yes they are the correct ones!

**Comment by camendola** [comment]
@francescobrivio everything is in now, you can review the changes!

File: interface/smallTree.h [hunk 2]
```diff
@@ -289,10 +366,18 @@ struct smallTree
...
+      m_HH_pt_raw_jetup.clear() ;
+      m_HH_pt_raw_taudown.clear() ;
+      m_HH_pt_raw_eledown.clear() ;
+      m_HH_pt_raw_jetdown.clear() ;
+      m_HH_mass_raw_tauup.clear() ;
+      m_HH_mass_raw_eleup.clear() ;
+      m_HH_mass_raw_jetup.clear() ;
+      m_HH_mass_raw_taudown.clear() ;
+      m_HH_mass_raw_eledown.clear() ;
+      m_HH_mass_raw_jetdown.clear() ;
```
Comment (francescobrivio) [inline interface/smallTree.h]: these can be removed

File: interface/smallTree.h [hunk 3]
```diff
@@ -177,35 +237,51 @@ struct smallTree
...
       m_bjet2_PUjetIDupdated = -99. ;
       m_bjet2_flav = -1 ;
       m_bjet2_pt_raw  = -1. ;
-      m_bjet2_pt_raw_jetup  = -1. ;
-      m_bjet2_pt_raw_jetdown  = -1. ;
+      m_bjet2_pt_raw_jetup.clear() ;
+      m_bjet2_pt_raw_jetdown.clear() ;
+      m_bjet2_mass_raw_jetup.clear() ;
+      m_bjet2_mass_raw_jetdown.clear() ;
+
```
Comment (francescobrivio) [inline interface/smallTree.h:273]: add also "flat quantities" for bjets

File: test/skimNtuple2017.cpp [hunk 4]
```diff
@@ -304,61 +304,312 @@ vector<int> findSubjetIdxs (unsigned int iFatJet, bigTree & theBigTree)
   return indexes;
 }
 
+// get shifted MET
+//returns a pair of vectors
+//first: up variations
+//second: down variations
+pair <vector <double>, vector <double> > getJetUpDown(int idx, int size, bigTree & theBigTree){
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp]: I think the parametere `int size` is never used in this function

File: test/skimNtuple2017.cpp [hunk 5]
```diff
@@ -304,61 +304,312 @@ vector<int> findSubjetIdxs (unsigned int iFatJet, bigTree & theBigTree)
...
+      //}
+      if (DEBUG) cout << "*********** ************** *********** "<< endl;
+    }
+  }
+  for (int isource = 0; isource < N_jecSources; isource++)
+  {
+    shiftedMET_up[isource].Set(corrMETx_up[isource], corrMETy_up[isource]);
+    shiftedMET_up[isource].Set(corrMETx_up[isource], corrMETy_up[isource]);
+    shiftedMET_down[isource].Set(corrMETx_down[isource], corrMETy_down[isource]);
+    shiftedMET_down[isource].Set(corrMETx_down[isource], corrMETy_down[isource]);
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp:419]: why are these two lines repeated?

File: test/skimNtuple2017.cpp [hunk 6]
```diff
@@ -1964,113 +2218,171 @@ int main (int argc, char** argv)
...
+          tlv_firstLepton_eleup[idm]   = getShiftedDau(tlv_firstLepton, 1.,  unc_EESup_first[idm], isthisDM_first[idm]);  
+          tlv_firstLepton_eledown[idm] = getShiftedDau(tlv_firstLepton, -1., unc_EESdw_first[idm], isthisDM_first[idm]);  
+        }
+        if (lep2HasTES)
+        {
+          tlv_secondLepton_tauup[idm]   = getShiftedDau(tlv_secondLepton, 1.,  unc_TES_second[idm], isthisDM_second[idm], (idm != 0));  // no mass shift for DM == 0 (idm == 0)
+          tlv_secondLepton_taudown[idm] = getShiftedDau(tlv_secondLepton, -1., unc_TES_second[idm], isthisDM_second[idm], (idm != 0));  // no mass shift for DM == 0 (idm == 0)
+        }else if(lep2HasEES && idm < N_tauhDM_EES){
+          tlv_secondLepton_tauup[idm]   = getShiftedDau(tlv_secondLepton, 1.,  unc_EESup_second[idm], isthisDM_second[idm]);  
+          tlv_secondLepton_taudown[idm] = getShiftedDau(tlv_secondLepton, -1., unc_EESdw_second[idm], isthisDM_second[idm]);
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp]: these should be `_eleup`and `_eledown` I think

File: test/skimNtuple2017.cpp [hunk 7]
```diff
@@ -1964,113 +2218,171 @@ int main (int argc, char** argv)
...
+    
+        if (idm < N_tauhDM_EES)
+        {
+          theSmallTree.m_dau1_pt_eleup.push_back(tlv_firstLepton_eleup[idm].Pt());
+          theSmallTree.m_dau1_pt_eledown.push_back(tlv_firstLepton_eledown[idm].Pt());
+          theSmallTree.m_dau1_mass_eleup.push_back(tlv_firstLepton_eleup[idm].M());
+          theSmallTree.m_dau1_mass_eledown.push_back(tlv_firstLepton_eledown[idm].M());
+          theSmallTree.m_dau2_pt_eleup.push_back(tlv_secondLepton_eleup[idm].Pt());
+          theSmallTree.m_dau2_pt_eledown.push_back(tlv_secondLepton_eledown[idm].Pt());
+          theSmallTree.m_dau2_mass_eleup.push_back(tlv_secondLepton_eleup[idm].M());
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp:2312]: `theSmallTree.m_dau2_mass_eledown` is missing

File: test/skimNtuple2017.cpp [hunk 8]
```diff
@@ -3411,8 +3825,19 @@ int main (int argc, char** argv)
...
+             
               theSmallTree.m_BDT_HT20 += tlv_jet.Pt() ;
+
               if (DEBUG) cout << " ---> Jet " << iJet << " - pt: " << tlv_jet.Pt() << " - HT: " << theSmallTree.m_BDT_HT20 << endl;
+              pair <vector <double>, vector<double>> unc_updown = getJetUpDown(iJet, N_jecSources, theBigTree);
+              // build shifted jet
+              for (int isource = 0; isource < N_jecSources; isource++)
+              {
+                TLorentzVector tlv_jetup   = getShiftedJet(tlv_jet, +1., unc_updown.first[isource]);
+                TLorentzVector tlv_jetdown = getShiftedJet(tlv_jet, -1., unc_updown.second[isource]);
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp:3838]: shouldn't we check here that these shifted jets have pT > 20 GeV?

File: test/skimNtuple2017.cpp [hunk 9]
```diff
@@ -3443,21 +3869,55 @@ int main (int argc, char** argv)
...
+        theSmallTree.m_METx_taudown.push_back(vMET_shift_tes.second.at(idm).X());
+        theSmallTree.m_METy_taudown.push_back(vMET_shift_tes.second.at(idm).Y());
+        if (idm <N_tauhDM_EES){
+          theSmallTree.m_METx_eleup.push_back(vMET_shift_ees.first.at(idm).X());
+          theSmallTree.m_METy_eleup.push_back(vMET_shift_ees.first.at(idm).Y());
+          theSmallTree.m_METx_eledown.push_back(vMET_shift_ees.second.at(idm).X());
+          theSmallTree.m_METy_eledown.push_back(vMET_shift_ees.second.at(idm).Y());
+        }
+      }
+       // Shifted MET for TES
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp]: this commented line should be `// Shifted MET for MES`

File: test/skimNtuple2017.cpp [hunk 10]
```diff
@@ -3519,7 +3968,7 @@ int main (int argc, char** argv)
 
         } // in case the SVFIT mass is calculated
 
-        // compute HHKinFit -- ask a reasonable mass window to suppress most error messages
+        // compute HHKinFit -- ask a ronable mass window to suppress most error messages
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp]: ??

File: test/skimNtuple2017.cpp [hunk 11]
```diff
@@ -3913,6 +4362,44 @@ int main (int argc, char** argv)
           TLorentzVector VBFjet2;
           VBFjet2.SetPxPyPzE( theBigTree.jets_px->at(VBFidx2), theBigTree.jets_py->at(VBFidx2), theBigTree.jets_pz->at(VBFidx2), theBigTree.jets_e->at(VBFidx2) );
 
+	  pair <vector <double>, vector<double>> unc_VBF1_updown = getJetUpDown(VBFidx1, N_jecSources, theBigTree);
+	  pair <vector <double>, vector<double>> unc_VBF2_updown = getJetUpDown(VBFidx2, N_jecSources, theBigTree);
+	  
+	  // compute all shifted bjets
```
Comment (francescobrivio) [inline test/skimNtuple2017.cpp]: commented line should be `// compute all shifted VBFjets`

File: test/skimNtuple2017.cpp (hunk 8 above)
Comment (camendola) [reply test/skimNtuple2017.cpp:3838]: uhm I'm not sure, but I can implement that

Comment (camendola) [reply test/skimNtuple2017.cpp:3838]: done, they all need to pass the PT > 20 GeV requirement

**Pull request merged**